src/test/java/be/syntra/devshop/DevshopBack/services/ProductServiceTest.java

```diff
@@ -268,4 +272,59 @@ void canFindProductRoundedPriceMaxTest(boolean archived) {
         assertThat(roundedResult).isEqualTo(price);
         verify(productRepository, times(1)).findRoundedMaxPrice(archived);
     }
+
+    @Test
+    void canSubmitReviewTest() {
+        // given
+        Product dummyProduct = createNonArchivedProduct();
+        final Review dummyReview = getDummyReview();
+        final Review dummyOtherReview = getDummyOtherReview();
+        Set<Review> reviewSet = new HashSet<>();
+        reviewSet.add(dummyOtherReview);
+        dummyProduct.setReviews(reviewSet);
+        when(productRepository.findById(dummyProduct.getId())).thenReturn(Optional.of(dummyProduct));
+        when(productRepository.save(any(Product.class))).thenReturn(dummyProduct);
+        // when
+        Product resultProduct = productService.submitReview(dummyReview, dummyProduct.getId());
+
+        // then
+        assertEquals(dummyProduct.getReviews(), resultProduct.getReviews());
+    }
+
+    @Test
+    void canRemoveReviewTest() {
+        // given
+        Product dummyProduct = createNonArchivedProduct();
+        final Review dummyReview = getDummyReview();
+        Set<Review> reviewSet = new HashSet<>();
+        reviewSet.add(dummyReview);
+        dummyProduct.setReviews(reviewSet);
+        when(productRepository.findById(dummyProduct.getId())).thenReturn(Optional.of(dummyProduct));
+        when(productRepository.save(any(Product.class))).thenReturn(dummyProduct);
+        // when
+        Product resultProduct = productService.removeReview(dummyReview, dummyProduct.getId());
+
+        // then
+        assertEquals(dummyProduct.getReviews(), resultProduct.getReviews());
+    }
+
+    @Test
+    void canUpdateReviewTest() {
+        // given
+        Product dummyProduct = createNonArchivedProduct();
+        Review dummyReview = getDummyReview();
+        Review dummyOtherReview = getDummyOtherReview();
+        dummyOtherReview.setUserName(dummyReview.getUserName());
+        Set<Review> reviewSet = new HashSet<>();
+        reviewSet.add(dummyReview);
+        dummyProduct.setReviews(reviewSet);
+        when(productRepository.findById(dummyProduct.getId())).thenReturn(Optional.of(dummyProduct));
+        when(productRepository.save(any(Product.class))).thenReturn(dummyProduct);
+        // when
+        Product resultProduct = productService.updateReview(dummyOtherReview, dummyProduct.getId());
+
+        // then
+        assertEquals(((Review) resultProduct.getReviews().toArray()[0]).getReviewText(), dummyOtherReview.getReviewText());
```

Comment:
a Set has a .contains() and .containsAll() method for checking the content

Reply:
because of the altered .equals of review we can't check the review itself, so we only check the reviewText property